tests/accelerators/test_tpu_backend.py

```diff
@@ -186,3 +190,77 @@ def test_set_devices_if_none_tpu():
 
     trainer = Trainer(accelerator="tpu", tpu_cores=8)
     assert trainer.devices == 8
+
+
+@RunIf(tpu=True)
+def test_manual_optimization_tpus(tmpdir):
+
+    class ManualOptimizationModel(BoringModel):
+
+        count = 0
+        called = collections.defaultdict(int)
+
+        def __init__(self):
+            super().__init__()
+            self.automatic_optimization = False
+
+        @property
+        def should_update(self):
+            return self.count % 2 == 0
+
+        def on_train_batch_start(self, batch, batch_idx, dataloader_idx):
+            self.called["on_train_batch_start"] += 1
+            self.weight_before = self.layer.weight.clone()
+
+        def training_step(self, batch, batch_idx):
+            self.called["training_step"] += 1
+            opt = self.optimizers()
+            output = self.layer(batch)
+            loss = self.loss(batch, output)
+
+            if self.should_update:
+                self.manual_backward(loss)
+                opt.step()
+                opt.zero_grad()
+            return loss
+
+        def on_train_batch_end(self, outputs, batch, batch_idx, dataloader_idx):
+            self.called["on_train_batch_end"] += 1
+            after_before = self.layer.weight.clone()
+            if self.should_update:
+                assert not torch.equal(self.weight_before, after_before), self.count
+            else:
+                assert torch.equal(self.weight_before, after_before)
+            assert torch.all(self.layer.weight.grad == 0)
+            self.count += 1
+
+        def on_train_end(self):
+            assert self.called["training_step"] == 5
+            assert self.called["on_train_batch_start"] == 5
+            assert self.called["on_train_batch_end"] == 5
+
+    class TestManualOptimizationCallack(Callback):
+
+        def on_train_end(self, trainer, pl_module):
+
+            opt = pl_module.optimizers()
+            assert opt._total_optimizer_step_calls == 3
```

Comment:
@kaushikb11 can we not count the optimizer step calls through mocking `torch.optim.Optimizer`? 
Otherwise we are relying on our internal counters

Reply:
Weirdly, mocking doesn't seem to be going well with training on 8 cores (spawn). Seems to be working with a single core.

Reply:
Hey @awaelchli, not sure TPU works nicely with mocking the optimizer.

Reply:
no problem, it would be nice to test this way but if there are limitations let's leave if then